app/exporters/patients_with_history_exporter.rb

```diff
@@ -0,0 +1,204 @@
+require "csv"
+
+module PatientsWithHistoryExporter
+  extend QuarterHelper
+
+  BATCH_SIZE = 20
+  BLOOD_SUGAR_TYPES = {
+    random: "Random",
+    post_prandial: "Postprandial",
+    fasting: "Fasting",
+    hba1c: "HbA1c"
+  }.with_indifferent_access.freeze
+  DISPLAY_BLOOD_PRESSURES = 3
+  DISPLAY_MEDICATION_COLUMNS = 5
+
+  def self.csv(patients)
+    CSV.generate(headers: true) do |csv|
+      csv << timestamp
+      csv << csv_headers
+
+      patients.in_batches(of: BATCH_SIZE).each do |batch|
+        batch.includes(
+          :registration_facility,
+          :phone_numbers,
+          :address,
+          :medical_history,
+          :latest_bp_passports,
+          {appointments: :facility},
+          {latest_blood_pressures: :facility},
+          :latest_blood_sugars
+        ).each do |patient|
+          csv << csv_fields(patient)
+        end
+      end
+    end
+  end
+
+  def self.timestamp
+    [
+      "Report generated at:",
+      Time.current
+    ]
+  end
+
+  def self.csv_headers
+    ["Registration Date",
+      "Registration Quarter",
+      "Patient died?",
+      "Simple Patient ID",
+      "BP Passport ID",
+      "Patient Name",
+      "Patient Age",
+      "Patient Gender",
+      "Patient Phone Number",
+      "Patient Street Address",
+      "Patient Village/Colony",
+      "Patient District",
+      (zone_column if Rails.application.config.country[:patient_line_list_show_zone]),
+      "Patient State",
+      "Registration Facility Name",
+      "Registration Facility Type",
+      "Registration Facility District",
+      "Registration Facility State",
+      "Risk Level",
+      "Days Overdue For Next Follow-up",
+      (1..DISPLAY_BLOOD_PRESSURES).map do |i|
+        ["BP #{i} Date",
+          "BP #{i} Quarter",
+          "BP #{i} Systolic",
+          "BP #{i} Diastolic",
+          "BP #{i} Facility Name",
+          "BP #{i} Facility Type",
+          "BP #{i} Facility District",
+          "BP #{i} Facility State",
+          "BP #{i} Follow-up Facility",
+          "BP #{i} Follow-up Date",
+          "BP #{i} Follow up Days",
+          "BP #{i} Medication Updated",
+          "BP #{i} Medication 1",
+          "BP #{i} Dosage 1",
+          "BP #{i} Medication 2",
+          "BP #{i} Dosage 2",
+          "BP #{i} Medication 3",
+          "BP #{i} Dosage 3",
+          "BP #{i} Medication 4",
+          "BP #{i} Dosage 4",
+          "BP #{i} Medication 5",
+          "BP #{i} Dosage 5",
+          "BP #{i} Other Medications"]
+      end,
+      "Latest Blood Sugar Date",
+      "Latest Blood Sugar Value",
+      "Latest Blood Sugar Type"].flatten.compact
+  end
+
+  def self.csv_fields(patient)
+    registration_facility = patient.registration_facility
+    latest_bps = patient.latest_blood_pressures.first(DISPLAY_BLOOD_PRESSURES)
```

Comment:
Don't we need to grab one more BP? How do we populate the "medication updated?" correctly on the oldest BP?

Reply:
Hrm that's correct. Right now it gets the medication before the oldest BP as `none` and compares to that. Thanks for spotting :eyes: